base/compiler/abstractinterpretation.jl

```diff
@@ -887,6 +887,12 @@ function abstract_call_method_with_const_args(interp::AbstractInterpreter,
     if !const_prop_enabled(interp, sv, match)
         return nothing
     end
+    if is_removable_if_unused(result.effects)
+        if isa(result.rt, Const) || call_result_unused(sv)
+            # There is no more information to be gained here. Bail out early.
+            return nothing
+        end
+    end
```

Comment:
Maybe we should implement this logic within `concrete_eval_eligible`? There might be a chance for const-prop' to shape up the optimized IR body even if it can't refine the type/effects information.

Reply:
The IR will not be inlined ideally. We should see it's dead and remove it

Reply:
If the IR body is inlineable but not simple enough for it to be folded into `ConstAPI`, it may hit this branch https://github.com/JuliaLang/julia/blob/8d9d23d2b063cdb3a07fb00ed6360b8620c995a7/base/compiler/ssair/inlining.jl#L1015-L1016
and end up being expanded into the callee.

Having said that, we intentionally miss this sort of optimization opportunity, like:
https://github.com/JuliaLang/julia/blob/86e0c6aad3860b9e41c518bb145ca1d8269b22aa/base/compiler/abstractinterpretation.jl#L1003-L1008

Reply:
I've thought about this some more, but I'm not sure it's really worth the cost to do the extra constprop or semiconcrete eval. Yes, it's possible that would shape up the IR farther, but we already have very strong effects on this. If a case comes up where this becomes an issue, I'd rather just see if there's something we can do to strengthen the optimizer.